torch/distributed/pipelining/stage.py

```diff
@@ -616,6 +616,15 @@ def perform_backward(
             else:
                 raise RuntimeError(f"Unknown backward type: {backward_type}")
 
+        def perform_pp_grad_scaling():
```

Comment:
@H-Huang any thoughts on the overall way i plumbed this from schedule to stage?

i thought about how schedule is the place where loss fn is registered and also the place that tracks `last_backward`, so if we want to expose a user-API it might make sense for that to be in PipelineSchedule class.

Reply:
Yeah I agree the user-API should be at the `PipelineSchedule` class as well.

If we do the pp_grad_scaling after DP reduction, then the scaling can be entirely at the schedule implementation instead of passing in the scale factor into stage. Do you think that would be a better place? (though we would have to figure out the numerics differences between scaling before or after you mentioned below)

Reply:
yea, it might be cleaner to move scaling into the schedule.  I was thinking at the time that it would require putting the scaling code in multiple schedule impls.  If we could do it cleanly, then i'd prefer that.

Reply:
i made this change. scaling is now called from the schedule.